### PR TITLE
LocalScanner: Check for archived files if scan results are present

### PR DESCRIPTION
**Pull request by KorSin**
The LocalScanner supports both an archive and a storage. So far, if
results of a package are present in the storage, they are skipped,
independent of the presence of results in the archive. Add an additional
check that, in case of absence, downloads the source code and applies the
archiving process.

### CODE REVIEW

File: scanner/src/main/kotlin/LocalScanner.kt [hunk 1]
```diff
@@ -260,6 +260,14 @@ abstract class LocalScanner(name: String, config: ScannerConfiguration) : Scanne
                             "and $scannerCriteria, not scanning the package again $packageIndex."
                 }
 
+                storedResults.forEach {
+                    if (!archiver.hasArchive(pkg.id.toPath() + "/" + it.provenance.hash())) {
```
Comment (mnonnenmacher) [inline scanner/src/main/kotlin/LocalScanner.kt:264]: `"${pkg.id.toPath()}/${it.provenance.hash()}"` for consistency. We rarely use string concatenation.

**Review by mnonnenmacher — DISMISSED** [review]

File: scanner/src/main/kotlin/LocalScanner.kt [hunk 2]
```diff
@@ -260,6 +260,14 @@ abstract class LocalScanner(name: String, config: ScannerConfiguration) : Scanne
                             "and $scannerCriteria, not scanning the package again $packageIndex."
                 }
 
+                storedResults.forEach {
+                    if (!archiver.hasArchive("${pkg.id.toPath()}/${it.provenance.hash()}")) {
+                        val downloadResult = Downloader.download(pkg, downloadDirectory.resolve(pkg.id.toPath()))
+
```
Comment (sschuberth) [inline scanner/src/main/kotlin/LocalScanner.kt:266]: Nit: Could drop this superfluous empty line.

File: scanner/src/main/kotlin/LocalScanner.kt [hunk 3]
```diff
@@ -260,6 +260,14 @@ abstract class LocalScanner(name: String, config: ScannerConfiguration) : Scanne
                             "and $scannerCriteria, not scanning the package again $packageIndex."
                 }
 
+                storedResults.forEach {
+                    if (!archiver.hasArchive("${pkg.id.toPath()}/${it.provenance.hash()}")) {
+                        val downloadResult = Downloader.download(pkg, downloadDirectory.resolve(pkg.id.toPath()))
```
Comment (sschuberth) [inline scanner/src/main/kotlin/LocalScanner.kt:265]: Nit: Could extract `pkg.id.toPath()` to a variable that is used here and above.

File: scanner/src/main/kotlin/LocalScanner.kt [hunk 4]
```diff
@@ -260,6 +260,14 @@ abstract class LocalScanner(name: String, config: ScannerConfiguration) : Scanne
                             "and $scannerCriteria, not scanning the package again $packageIndex."
                 }
 
+                storedResults.forEach {
+                    if (!archiver.hasArchive("${pkg.id.toPath()}/${it.provenance.hash()}")) {
```
Comment (KorSin) [inline scanner/src/main/kotlin/LocalScanner.kt:264]: At the moment, I am a bit unsure about the case when the `storedResults` have more than one entry. The current code would download the same(?) source code each time and archive the result for each provenance. The download should obviously be called only once, that I need to change. But does archiving the same results for each provenance make sense?

It seems that the provenance  depends on download date and VCS data. Does this mean that the corresponding hash depends on the state of the repository when the package was first downloaded and scanned? If so, should the current archive result be put together with an earlier scan result? I am thinking about packages which are continuously updated and have no real versioning.

Maybe someone can enlighten me.

Comment (mnonnenmacher) [reply scanner/src/main/kotlin/LocalScanner.kt:264]: I didn't think of the case before that there could be multiple results for the same provenance, e.g. from different scanners. I think it would be better to first collect all `Provenance` objects in a set and then iterate over this to prevent downloading the same source multiple times.

Another case would be multiple results with different provenance, e.g. one for VCS and one for a source artifact. In this case it would be correct to download it multiple times, as the archived files could be different.

File: scanner/src/main/kotlin/LocalScanner.kt [hunk 5]
```diff
@@ -260,6 +260,22 @@ abstract class LocalScanner(name: String, config: ScannerConfiguration) : Scanne
                             "and $scannerCriteria, not scanning the package again $packageIndex."
                 }
 
+                val packagePath = pkg.id.toPath()
+                val unarchivedProvenances = HashSet<Provenance>()
+
+                storedResults.forEach {
+                    if (!archiver.hasArchive("$packagePath/${it.provenance.hash()}")) {
+                        unarchivedProvenances.add(it.provenance)
+                    }
```
Comment (mnonnenmacher) [inline scanner/src/main/kotlin/LocalScanner.kt:269]: Did you check how long this takes for a large scan with several thousand dependencies? If it is more than a few seconds consider wrapping the loop with `measureTime` and adding a performance log with `log.perf`.

Comment (mnonnenmacher) [reply scanner/src/main/kotlin/LocalScanner.kt:269]: Thanks for checking, 30s doesn't sound like much, but if you run scans in a CI/CD environment it can be relevant. Usually you only need this check if you changed the location of the artifact storage or if you use scan results from an external provider, so this check is not always required. I'm just thinking if it would be better to add a command line option to the scanner to enable/disable this check. What do you think @sschuberth @fviernau ?

File: scanner/src/main/kotlin/LocalScanner.kt [hunk 6]
```diff
@@ -260,6 +260,22 @@ abstract class LocalScanner(name: String, config: ScannerConfiguration) : Scanne
                             "and $scannerCriteria, not scanning the package again $packageIndex."
                 }
 
+                val packagePath = pkg.id.toPath()
```
Comment (sschuberth) [inline scanner/src/main/kotlin/LocalScanner.kt:263]: I read through the commit message again and propose the following changes:

> LocalScanner: Check for archived files if scan results are present

This does not only check for presence, but also recreate the archives if needed. So I'd change this to "LocalScanner: (Re-)Create license file archives if needed"

The body text I'd reword as follows:

The LocalScanner supports both an archive (for license files) and a
storage (for scan results) for a package. So far, if scan results are
present in the storage, scanning is skipped independently of the presence
of an archive. Allow to (re-)create the archive in such cases so reporters
depending on the archive work properly.

File: scanner/src/main/kotlin/LocalScanner.kt [hunk 7]
```diff
@@ -260,6 +260,22 @@ abstract class LocalScanner(name: String, config: ScannerConfiguration) : Scanne
...
+                val unarchivedProvenances = HashSet<Provenance>()
```
Comment (sschuberth) [inline scanner/src/main/kotlin/LocalScanner.kt:264]: I'd name this just `missingArchives`.

File: scanner/src/main/kotlin/LocalScanner.kt [hunk 8]
```diff
@@ -260,6 +260,22 @@ abstract class LocalScanner(name: String, config: ScannerConfiguration) : Scanne
...
+
+                storedResults.forEach {
+                    if (!archiver.hasArchive("$packagePath/${it.provenance.hash()}")) {
+                        unarchivedProvenances.add(it.provenance)
+                    }
+                }
+
+                if (unarchivedProvenances.size > 0) {
```
Comment (sschuberth) [inline scanner/src/main/kotlin/LocalScanner.kt:272]: Use `.isNotEmpty()`.

File: scanner/src/main/kotlin/LocalScanner.kt (hunk 5 above)
Comment (sschuberth) [reply scanner/src/main/kotlin/LocalScanner.kt:269]: Yes, let's add a scanner option for this that is exposed as a CLI flag and defaults to `false`. I propose a flag name of `--create-missing-archives`.

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: cli/src/main/kotlin/commands/ScannerCommand.kt [hunk 9]
```diff
@@ -118,6 +118,11 @@ class ScannerCommand : CliktCommand(name = "scan", help = "Run existing copyrigh
         help = "Do not scan excluded projects or packages. Works only with the '--ort-file' parameter."
     ).flag()
 
+    private val createMissingArchives by option(
+        "--create-missing-archives",
+        help = "Creates missing archives for packages even if scan results are present in the storage."
```
Comment (sschuberth) [inline cli/src/main/kotlin/commands/ScannerCommand.kt:123]: Use the imperative: s/Create/Creates/

And I'd say "Create archives for packages that have a stored scan result but no license archive yet."

Comment (sschuberth) [reply cli/src/main/kotlin/commands/ScannerCommand.kt:123]: Thinking about this again, @mnonnenmacher should we make this a scanner configuration option instead of a CLI option, in our effort to remove configuration options from the CLI?

Comment (mnonnenmacher) [reply cli/src/main/kotlin/commands/ScannerCommand.kt:123]: Makes sense to me as it could still be passed on the command line if it's only required for a single run.

File: scanner/src/main/kotlin/LocalScanner.kt [hunk 10]
```diff
@@ -269,6 +272,24 @@ abstract class LocalScanner(name: String, config: ScannerConfiguration) : Scanne
...
 
+                if (createMissingArchives) {
+                    val packagePath = pkg.id.toPath()
+                    val missingArchives = HashSet<Provenance>()
+
+                    storedResults.forEach {
+                        if (!archiver.hasArchive("$packagePath/${it.provenance.hash()}")) {
+                            missingArchives.add(it.provenance)
+                        }
+                    }
```
Comment (sschuberth) [inline scanner/src/main/kotlin/LocalScanner.kt:283]: Could be written as

                    val missingArchives = storedResults.mapNotNullTo(mutableSetOf()) { result ->
                        result.provenance.takeUnless { archiver.hasArchive("$packagePath/${result.provenance.hash()}") }
                    }

File: cli/src/main/kotlin/commands/ScannerCommand.kt (hunk 9 above)
Comment (KorSin) [reply cli/src/main/kotlin/commands/ScannerCommand.kt:123]: Would you rather have it in the `ScannerConfiguration` or in the `FileArchiverConfiguration`? I cannot decide.

Comment (KorSin) [reply cli/src/main/kotlin/commands/ScannerCommand.kt]: I put it inside `FileArchiverConfiguration`.

Comment (sschuberth) [reply cli/src/main/kotlin/commands/ScannerCommand.kt]: Sorry for replying (too) late... I would have put it in the `ScannerConfiguration`, as it's the scanner, not the archiver, who takes the configuration into account. I would have put it just below `val archive: FileArchiverConfiguration? = null`.

File: scanner/src/main/kotlin/LocalScanner.kt [hunk 11]
```diff
@@ -104,6 +104,10 @@ abstract class LocalScanner(name: String, config: ScannerConfiguration) : Scanne
         config.archive.createFileArchiver()
     }
 
+    private val createMissingArchives by lazy {
+        config.archive?.createMissingArchives ?: false
```
Comment (sschuberth) [inline scanner/src/main/kotlin/LocalScanner.kt:108]: Getting this property is very cheap, so there should be no need to evaluate it lazily.

**Comment by sschuberth** [comment]
And please rebase onto latest `master` to fix CI checks.

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: model/src/main/kotlin/config/FileArchiverConfiguration.kt [hunk 12]
```diff
@@ -33,7 +33,7 @@ data class FileArchiverConfiguration(
     /**
      * Configuration of the [FileStorage] used for archiving the files.
      */
-    val storage: FileStorageConfiguration
+    val storage: FileStorageConfiguration,
```
Comment (sschuberth) [inline model/src/main/kotlin/config/FileArchiverConfiguration.kt:36]: Please create the left-over trailing comma. (Even if it's actually valid Kotlin syntax!)

File: model/src/test/assets/reference.conf [hunk 13]
```diff
@@ -39,6 +39,7 @@ ort {
           compression = false
         }
       }
+      createMissingArchives = false
```
Comment (sschuberth) [inline model/src/test/assets/reference.conf:42]: Nit: Blank line before this one.

Comment (sschuberth) [reply model/src/test/assets/reference.conf:42]: And this should be on the same level as `archive` now.

File: scanner/src/main/kotlin/LocalScanner.kt [hunk 14]
```diff
@@ -104,6 +104,8 @@ abstract class LocalScanner(name: String, config: ScannerConfiguration) : Scanne
         config.archive.createFileArchiver()
     }
 
+    private val createMissingArchives = config.createMissingArchives
```
Comment (sschuberth) [inline scanner/src/main/kotlin/LocalScanner.kt:107]: Let's just inline this to get rid of the variable altogether.

**Review by sschuberth — CHANGES_REQUESTED** [review]

**Comment by sschuberth** [comment]
You still need to add the new config to `file-counter-expected-output-for-analyzer-result.yml` to make tests pass.

**Review by sschuberth — APPROVED** [review]
